This patch is a combination of

commit eecf592b0b2373b6c9901de31723c4caf230387f
Author: ihayvuk <ihayvuk>
Date:   Wed Aug 18 09:43:48 2010 +0000

    Fixed bug #1521 (LabelGadget draws the text over border of its parent)

commit d6f409ef6778ea68fdc5362e4eef46ddb7ebcfb5
Author: ihayvuk <ihayvuk>
Date:   Thu Nov 11 09:47:12 2010 +0000

    Added missing #ifdef USE_XFT directive to the patch #1521

The fix for the regression introduced by patch #1521 (rhbz#729346):
commit 117a3d65431d75f31824252ab97352e8f2e0541f
Author: ihayvuk <ihayvuk>
Date:   Wed Aug 10 14:15:10 2011 +0000

    Added missing include

Fir for Label draws the text over border of its parent:
commit 2f574ed86a3fdaac9a23124f6147a4b60b42dec6
Author: ihayvuk <ihayvuk>
Date:   Mon Aug 15 13:31:47 2011 +0000

    Refixed bug #1521 (LabelGadget draws the text over border of its parent)

diff -up openmotif-2.3.3/lib/Xm/LabelG.c.motifzone_1521 openmotif-2.3.3/lib/Xm/LabelG.c
--- openmotif-2.3.3/lib/Xm/LabelG.c.motifzone_1521	2011-08-01 15:32:32.508243633 +0200
+++ openmotif-2.3.3/lib/Xm/LabelG.c	2011-08-01 15:33:45.522396184 +0200
@@ -97,6 +97,11 @@ static char rcsid[] = "$TOG: LabelG.c /m
 #ifdef FIX_1381
 #include <Xm/ColorI.h>
 #endif
+#ifdef FIX_1521
+#ifdef USE_XFT
+#include <X11/Xft/Xft.h>
+#endif
+#endif
 
 #define Pix(w)      LabG_Pixmap(w)
 #define Pix_insen(w)    LabG_PixmapInsensitive(w)
@@ -2332,8 +2337,21 @@ LRectangle *background_box)
             clip_rect.height = 0;
 
         XSetClipRectangles(XtDisplay(lw), clipgc, 0,0, &clip_rect, 1, Unsorted);
+#ifdef FIX_1521
+#ifdef USE_XFT
+        _XmXftSetClipRectangles(XtDisplay(lw), XtWindow(lw), 0, 0, &clip_rect, 1);
+#endif
+#endif
     } else
+    {
     XSetClipMask (XtDisplay (lw), clipgc, None);
+#ifdef FIX_1521
+#ifdef USE_XFT
+	XftDraw	*draw = _XmXftDrawCreate(XtDisplay(lw), XtWindow(lw));
+	XftDrawSetClip(draw, NULL);
+#endif
+#endif
+    }
 
 #ifdef FIX_1517
 #ifdef USE_XFT
diff -up openmotif-2.3.3/lib/Xm/XmI.h.motifzone_1521 openmotif-2.3.3/lib/Xm/XmI.h
--- openmotif-2.3.3/lib/Xm/XmI.h.motifzone_1521	2009-10-27 16:43:28.000000000 +0100
+++ openmotif-2.3.3/lib/Xm/XmI.h	2011-08-01 15:33:45.477396090 +0200
@@ -307,6 +307,7 @@ extern Pixel _XmAssignInsensitiveColor(W
 #define FIX_1402
 #define FIX_1445
 #define FIX_1474
+#define FIX_1521
 
 #endif /* _XmI_h */
 /* DON'T ADD ANYTHING AFTER THIS #endif */
diff -up openmotif-2.3.3/lib/Xm/XmString.c.motifzone_1521 openmotif-2.3.3/lib/Xm/XmString.c
--- openmotif-2.3.3/lib/Xm/XmString.c.motifzone_1521	2009-03-04 18:31:28.000000000 +0100
+++ openmotif-2.3.3/lib/Xm/XmString.c	2011-08-01 15:33:45.485396106 +0200
@@ -5097,7 +5097,11 @@ _render(Display *d,
             draw_x = base_x ; /* most left position */
             _calc_align_and_clip( d, w, gc, &draw_x, y, width, line_width, 
                                 lay_dir, clip, align, descender, 
+#ifdef FIX_1521
+                                &restore_clip, _XmRendFontType(rend2));
+#else
                                 &restore_clip, _XmRendFontType(rend));
+#endif
 
             DrawLine(d, w, &screen, draw_x, y, (_XmStringEntry)string, 
 		     &rend2, rend, rendertable, lay_dir, image, 
@@ -5127,7 +5131,11 @@ _render(Display *d,
 	  }
 	
 	/* width, height, ascent, descent of this line */
+#ifdef FIX_1521
+	LineMetrics(line, rendertable, &rend1, rend, lay_dir,
+#else
 	LineMetrics(line, rendertable, &rend2, rend, lay_dir,
+#endif
 		    &line_width, &line_height, &ascender, &descender);
 
 	y += ascender;
@@ -5138,7 +5146,11 @@ _render(Display *d,
 
 	    _calc_align_and_clip(d, w, gc, &draw_x, y, width, line_width, 
 				 direction, clip, align, descender, 
+#ifdef FIX_1521
+				 &restore_clip, _XmRendFontType(rend1));
+#else
 				 &restore_clip, _XmRendFontType(rend));
+#endif
 
 	    DrawLine(d, w, &screen, draw_x, y, line, &rend1, rend,
 		     rendertable, lay_dir, image, &underline,
@@ -5153,7 +5165,17 @@ _render(Display *d,
 	y += descender;			  /* go to bottom of this line */
       }
   }
-  if (restore_clip) XSetClipMask (d, gc, None); 
+  if (restore_clip) {
+#ifdef FIX_1521
+#ifdef USE_XFT
+	  if (_XmRendFontType((_XmStrOptimized(string)) ? rend2 : rend1) == XmFONT_IS_XFT) {
+		  XftDraw *draw = _XmXftDrawCreate(d, w);
+		  XftDrawSetClip(draw, NULL);
+	  } else
+#endif
+#endif
+		  XSetClipMask (d, gc, None);
+  }
 
   if (_XmRendTags(rend1) != NULL) XtFree((char *)_XmRendTags(rend1));
   if (_XmRendTags(rend2) != NULL) XtFree((char *)_XmRendTags(rend2));
diff --git a/lib/Xm/LabelG.c b/lib/Xm/LabelG.c
index 1568c7f..091b076 100644
--- a/lib/Xm/LabelG.c
+++ b/lib/Xm/LabelG.c
@@ -99,6 +99,7 @@ static char rcsid[] = "$TOG: LabelG.c /main/24 1999/01/26 15:31:18 mgreess $"
 #endif
 #ifdef FIX_1521
 #ifdef USE_XFT
+#include "XmRenderTI.h"
 #include <X11/Xft/Xft.h>
 #endif
 #endif
diff --git a/lib/Xm/Label.c b/lib/Xm/Label.c
index 78ac19c..74853d7 100644
--- a/lib/Xm/Label.c
+++ b/lib/Xm/Label.c
@@ -89,6 +89,12 @@ static char rcsid[] = "$TOG: Label.c /main/26 1997/06/18 17:40:00 samborn $"
 #ifdef FIX_1381
 #include <Xm/ColorI.h>
 #endif
+#ifdef FIX_1521
+#ifdef USE_XFT
+#include "XmRenderTI.h"
+#include <X11/Xft/Xft.h>
+#endif
+#endif
 
 #define FIX_1442
 #define FIX_1484
@@ -1461,8 +1467,21 @@ Redisplay(
 	clip_rect.height = 0;
       
       XSetClipRectangles(XtDisplay(lw), clipgc, 0,0, &clip_rect, 1, Unsorted);
+#ifdef FIX_1521
+#ifdef USE_XFT
+      _XmXftSetClipRectangles(XtDisplay(lw), XtWindow(lw), 0, 0, &clip_rect, 1);
+#endif
+#endif
     } else
+    {
       XSetClipMask (XtDisplay (lw), clipgc, None);
+#ifdef FIX_1521
+#ifdef USE_XFT
+      XftDraw* draw = _XmXftDrawCreate(XtDisplay(lw), XtWindow(lw));
+      XftDrawSetClip(draw, NULL);
+#endif
+#endif
+    }
 
 #ifdef USE_XFT
     /* it is needed to clear anti-aliased text before draw it again */
diff --git a/lib/Xm/RCLayout.c b/lib/Xm/RCLayout.c
index b3951c0..19ac72c 100644
--- a/lib/Xm/RCLayout.c
+++ b/lib/Xm/RCLayout.c
@@ -1312,6 +1312,16 @@ LayoutHorizontaltight(
 	SetAsking (m, m_width, m_height, Double(border), 
 		    max_x, 0, x, y, w, h);
     
+#ifdef FIX_1521
+    for (child_i = 0; kg[child_i].kid != NULL; child_i++)
+    {
+        bx = &(kg[child_i].box);
+        ASSIGN_MIN(BWidth(bx), *m_width - Double(MGR_ShadowThickness(m))
+            - Double(RC_MarginW(m)) - Double(BBorder(bx)));
+        if (IsHelp(m, ((Widget) kg[child_i].kid)))
+            ASSIGN_MAX(BX(&(kg[child_i].box)), MGR_ShadowThickness(m) + RC_MarginW(m));
+    }
+#endif
     if (RC_AdjLast(m))
 	AdjustLast (m, start_i, *m_width, *m_height);
     else
